lib/Backend/Lower.cpp

```diff
@@ -3894,6 +4031,111 @@ Lowerer::GenerateProfiledNewScObjArrayFastPath(IR::Instr *instr, Js::ArrayCallSi
+        headOpnd = GenerateArrayAlloc<Js::JavascriptNativeIntArray>(instr, lengthOpnd, arrayInfo, &isZeroed, &alignedHeadSegSizeOpnd);
+
+        GenerateMemInit(dstOpnd, Js::JavascriptNativeIntArray::GetOffsetOfArrayCallSiteIndex(), IR::IntConstOpnd::New(profileId, TyUint16, func, true), instr, isZeroed);
+        GenerateMemInit(dstOpnd, Js::JavascriptNativeIntArray::GetOffsetOfWeakFuncRef(), IR::AddrOpnd::New(weakFuncRef, IR::AddrOpndKindDynamicFunctionBodyWeakRef, m_func), instr, isZeroed);
+    }
+    else if (arrayInfo && arrayInfo->IsNativeFloatArray())
+    {
+        elementSizeOpnd = IR::IntConstOpnd::New(sizeof(Js::JavascriptNativeFloatArray::TElement), TyUint8, func);
+        missingItem = GetMissingItemOpnd(IRType::TyVar, func);
+        GenerateArrayInfoIsNativeFloatAndNotIntArrayTest(instr, arrayInfo, helperLabel);
+        lengthOpnd = generateRangeCheck();
+        Assert(Js::JavascriptNativeFloatArray::GetOffsetOfArrayFlags() + sizeof(uint16) == Js::JavascriptNativeFloatArray::GetOffsetOfArrayCallSiteIndex());
+        headOpnd = GenerateArrayAlloc<Js::JavascriptNativeFloatArray>(instr, lengthOpnd, arrayInfo, &isZeroed, &alignedHeadSegSizeOpnd);
+
+        GenerateMemInit(dstOpnd, Js::JavascriptNativeFloatArray::GetOffsetOfArrayCallSiteIndex(), IR::IntConstOpnd::New(profileId, TyUint16, func, true), instr, isZeroed);
+        GenerateMemInit(dstOpnd, Js::JavascriptNativeFloatArray::GetOffsetOfWeakFuncRef(), IR::AddrOpnd::New(weakFuncRef, IR::AddrOpndKindDynamicFunctionBodyWeakRef, m_func), instr, isZeroed);
+    }
+    else
+    {
+        elementSizeOpnd = IR::IntConstOpnd::New(sizeof(Js::JavascriptArray::TElement), TyUint8, func);
+        missingItem = GetMissingItemOpnd(IRType::TyVar, func);
+        lengthOpnd = generateRangeCheck();
+        headOpnd = GenerateArrayAlloc<Js::JavascriptArray>(instr, lengthOpnd, arrayInfo, &isZeroed, &alignedHeadSegSizeOpnd);
+    }
+    //        MOV s13,          0                         ; index
+    //        MOV s14,          16                        ; headSegmentIndex
+
+    //  ; Need to hoist missingItem outsude loop else register allocator might reuse the same register to save the hoisted missingItem
+    //  ; as that of loopIndex since missingItem is not marked live on back edge.
+    //        MOV s15,          0x80000002                ; missingItem
+
+    //$LoopTop:
+    //        CMP s13,          alignedHeadSegSizeOpnd
+    //        JGE $done
+    //        MOV [s12+s14],    0x80000002                ; Save missing element
+    //        s14               = ADD s14, 4              ; headSegmentIndex += 4. NOTE: value 4 could be different for different type of arrays
+    //        s13               = ADD s13, 1              ; index ++
+    //        JMP $LoopTop
+    //
+    //$done
+
+    IR::RegOpnd * indexOpnd = IR::RegOpnd::New(TyUint32, func);
```

Comment:
> If we are limiting this fast path to only <= 8 (above in generateRangeCheck), 

Once this change goes through, I was thinking of bumping the restriction to 16 instead of 8.

> Can't we generate the code as if the user has asked for a constant 8 element, 

There is a special case where if user specified value is 0, we would just initialize 4 elements (`Js::SparseArraySegmentBase::SMALL_CHUNK_SIZE`) instead of 8.